homeassistant/components/p1_monitor/__init__.py

```diff
@@ -0,0 +1,91 @@
+"""The P1 Monitor integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+
+from p1monitor import P1Monitor, Phases, Settings, SmartMeter
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    CONF_TIME_BETWEEN_UPDATE,
+    DEFAULT_TIME_BETWEEN_UPDATE,
+    DOMAIN,
+    LOGGER,
+    SERVICE_PHASES,
+    SERVICE_SETTINGS,
+    SERVICE_SMARTMETER,
+)
+
+PLATFORMS = (SENSOR_DOMAIN,)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up P1 Monitor from a config entry."""
+    hass.data.setdefault(DOMAIN, {}).setdefault(entry.entry_id, {})
+
+    client = P1Monitor(host=entry.data[CONF_HOST])
+
+    min_time_between_updates = timedelta(
+        seconds=entry.options.get(CONF_TIME_BETWEEN_UPDATE, DEFAULT_TIME_BETWEEN_UPDATE)
+    )
+
+    async def update_smartmeter() -> SmartMeter:
+        return await client.smartmeter()
+
+    smartmeter: DataUpdateCoordinator[SmartMeter] = DataUpdateCoordinator(
+        hass,
+        LOGGER,
+        name=f"{DOMAIN}_{SERVICE_SMARTMETER}",
+        update_interval=min_time_between_updates,
+        update_method=update_smartmeter,
+    )
+    await smartmeter.async_config_entry_first_refresh()
```

Comment:
This call will potentially raise a ConfigEntryNotReady exception. If it does you currently leak the session that is allocated inside the client.

Reply:
Does this still apply since a lot has changed in the file?

Reply:
It can still throw an exception yes. At which point you "leak" a client. I say "leak" since as martin points out, that with the shared client session, the close() call is a noop. But since the API in the client library expects a close call, at some stage with perhaps a library bump , it may need further cleanup.

So i do think you should wrap that call in:
try:
  coordinator.first_update()
except ConfigEntryNotReady
  client.close()
  raise

Reply:
Latest commit will add this exception